src/main/java/org/candlepin/resource/util/EntitlementEnvironmentFilter.java

```diff
@@ -0,0 +1,147 @@
+        this.preExistingEnvs = preExistingEnvs;
+        return this;
+    }
+
+    public EntitlementEnvironmentFilter setUpdatedEnvironment(List<String> updatedEnvs) {
+        this.updatedEnvs = updatedEnvs;
+        return this;
+    }
+
+    public EntitlementEnvironmentFilter setConsumerToBeUpdated(Consumer consumer) {
+        this.consumer = consumer;
+        return this;
+    }
+
+    /**
+     * Returns the list of environment IDs which got added, removed or re-prioritized
+     * from consumers current existing environments.
+     *
+     * @return
+     *  Returns the unique environmentIds which got added, removed or re-prioritized
+     */
+    private List<String> getEnvironmentChangeList() {
+        Set<String> envChangeList = new HashSet<>();
+
+        // Envs Added
+        envChangeList.addAll(CollectionUtils.subtract(this.updatedEnvs, this.preExistingEnvs));
+
+        // Envs removed
+        envChangeList.addAll(CollectionUtils.subtract(this.preExistingEnvs, this.updatedEnvs));
+
+        // Envs reordered (Index within list changes)
+        for (String env : this.updatedEnvs) {
+            if (this.preExistingEnvs.indexOf(env) != this.updatedEnvs.indexOf(env)) {
+                envChangeList.add(env);
+            }
+        }
+
+        return new ArrayList<>(envChangeList);
+    }
+
+    /**
+     * Filters the entitlements who's certificates are needed to be re-generated due to environments
+     * being added, removed or re-prioritized for a consumer.
+     * Idea is to avoid regenerating all of the consumer's entitlements certs & filter the
+     * necessary entitlements who's certificates are actually needed to be re-generated.
+     *
+     * @return
+     *  List of entitlement selected for certificate regeneration
+     */
+    public List<Entitlement> filterEntitlements() {
+        List<Entitlement> entitlements = new ArrayList<>();
+        List<String> environmentChangeList = getEnvironmentChangeList();
+
+        for (Entitlement e : consumer.getEntitlements()) {
+            boolean entToRegen = false;
+            List<String> contentUUIDs = this.entitlementCurator
+                .getProductContentUUIDsByEntitlementId(e.getId());
+
+            for (String environmentId : environmentChangeList) {
+                List<String> currentHigher = new ArrayList();
+                Set<String> envListByCurrentHigher = new HashSet<>();
+                List<String> environmentContentUUIDs = this.environmentCurator
+                    .getEnvironmentContentUUIDs(environmentId);
+
+                if (this.updatedEnvs.contains(environmentId)) {
+                    currentHigher.addAll(this.updatedEnvs
+                        .subList(0, this.updatedEnvs.indexOf(environmentId)));
+                }
+                else {
+                    // Case where this environment is being deleted
+                    currentHigher.addAll(this.updatedEnvs);
+                }
```

Comment:
This will work in specific sub-cases of when the environment is being deleted, but I think I found an issue with the original algorithm implementation. I will have another go at it to find what exactly needs done for the rest of deletion sub-cases

Reply:
Fixed.